nixos/users-groups: do not check validity of special hashes

rnhmjoj opened the pull request:
###### Motivation for this change
Fixup of PR #83171

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- [x] Built manual
- [x] Tested via one or more NixOS test(s): `nixosTests.mutableUsers``
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @tilpner @erictapen

rnhmjoj committed:
nixos/users-groups: remove ancient security.initialRootPassword option

This option has been deprecated for a long time because is redundant
(users.users.root.initialHashedPassword exists).
Moreover, being of type string, it required to handle the special value
"!" separately, instead of using just `null`.

rnhmjoj committed:
nixos/users-groups: clearly document special hash values

This explanation was contained in the description of
security.initialRootPassword but got lost when it was deprecated
a long ago (f496c3c) and removed.

rnhmjoj committed:
nixos/users-groups: do not check validity of empty hashes

Profpatsch commented on nixos/modules/config/users-groups.nix:
```diff
@@ -406,6 +414,12 @@ in {
+      [ "users" "users" "root" "initialHashedPassword" ]
+      (cfg: if cfg.security.initialHashedPassword == "!"
+            then null
+            else cfg.security.initialHashedPassword))
```
This means that the change is backwards-compatible with people setting `security.initialRootPassword`, right?

rnhmjoj replied:
Yes, it is. Also this option has been hidden since 2014, so the impact should be small.

Profpatsch commented on nixos/modules/config/users-groups.nix:
```diff
@@ -588,7 +593,7 @@ in {
              || elem "wheel" cfg.extraGroups)
             &&
-            ((cfg.hashedPassword != null && cfg.hashedPassword != "!")
+            (cfg.hashedPassword != null
```
I’m super uncomfortable with this line, somebody else will have to confirm that it’s valid.

rnhmjoj replied:
There are two possible cases I can think of:
- using the deprecated `security.initialRootPassword` "!" will be turned into a null, so it's covered
- using `users.users.<name>.hashedPassword` or similar, "!" will issue the warning about possibly being locked out.

But yes, it would be better is someone familiar with this module would comment.

Profpatsch replied:
Looks to me like @edolstra is the only one who touched this code in the past.

jtojnar replied:
Yeah, user using `users.users.<name>.hashedPassword` that is not a hash will lock themselves out if they miss the warning so the assertion still makes sense. It is even too narrow here – I use `*` as a root’s hash and this would not save me. Not sure we can extend it much further without causing false positives, though.

jtojnar reviewed (commented):
1. We should definitely add an assertion that `hashedPassword` cannot contain `:` since that will definitely break the shadowfile as reported in #12008
2. There are other commonly used schemes for `hashedPassword`. Personally, I set it to `*` for root, since hash prefixed with `!` means the account is locked according to [`shadow(5)`](https://linux.die.net/man/5/shadow). Different systems might support special sequences with even wilder semantics like `!!` or `*NS*` https://unix.stackexchange.com/questions/252016/difference-between-vs-vs-in-etc-shadow

rnhmjoj commented:
> There are other commonly used schemes for hashedPassword. Personally, I set it to * for root, since hash prefixed with ! means the account is locked according to shadow(5). Different systems might support special sequences with even wilder semantics like !! or *NS*

What is the point of setting the hash to "!" when `null` is the default (and the `update-users-groups.pl` script locks users with "!" when the hash is `null`)? It's not a rhetoric question, I'm honestly asking.

If these special hashes are common I'll add exceptions, but to me it feel like an implementation detail that shouldn't leak to the NixOS configuration: the options `users.users` don't mention /etc/shadow anywhere.

jtojnar commented:
> What is the point of setting the hash to "!" when `null` is the default (and the `update-users-groups.pl` script locks users with "!" when the hash is `null`)? It's not a rhetoric question, I'm honestly asking.

I did not know that before reviewing the code. I would expect `null` to map to empty string (i.e. no password) so I set the hash to `*`. Also reading the `shadow(5)`, using `*` makes more sense to me than `!`.

> If these special hashes are common I'll add exceptions, but to me it feel like an implementation detail that shouldn't leak to the NixOS configuration: the options `users.users` don't mention /etc/shadow anywhere.

Yeah, this it a personal preference. Sometimes I like it when NixOS modules map directly to UNIX concepts and do not add much special logic. Especially, if I was familiar with the config before switching to NixOS.

rnhmjoj committed:
nixos/users-groups: add assertion for ":" in hashes

rnhmjoj committed:
nixos/users-groups: handle password hashes with special meaning

rnhmjoj commented:
I added a check for "!", "!!" and "*" in the assertion and the warning as you requested.
I also added another assertion for avoiding ":"s in the password hash, since these are guaranteed to break havoc.
The main concerns should be answered.

Profpatsch commented on nixos/modules/config/users-groups.nix:
```diff
@@ -603,7 +603,16 @@ in {
+                    -> (builtins.match ".*:.*" user.hashedPassword == null);
+        message = ''
+          The password hash of user "${name}" contains a ":" character.
+          This is invalid and would break the login system.
```
```suggestion
          This is invalid and would break the login system
          because /etc/passwd separates its fields with ":".
```

rnhmjoj replied:
Added something like it.

Profpatsch commented on nixos/modules/config/users-groups.nix:
```diff
@@ -6,6 +6,9 @@ let
   cfg = config.users;
 
+  # Check whether a password hash will allow login.
+  allowsLogin = hash: !(lib.elem hash [ null "*" "!" "!!" ]);
```
I’d add a comment to each form.

rnhmjoj replied:
Done.

Profpatsch commented on nixos/modules/config/users-groups.nix:
```diff
@@ -635,15 +638,13 @@ in {
-        if (user.hashedPassword != null   # login disabled
+        if (allowsLogin user.hashedPassword
             && user.hashedPassword != ""  # login without password
             && builtins.match mcf user.hashedPassword == null)
```
I think you already do these two checks with `allowsLogin`.

rnhmjoj replied:
If you mean the empty string, it's different: it's a special value but it does allow login, so it's not in `allowsLogin`.

Profpatsch replied:
So It should be in the `allowsLogin` check, right?

Profpatsch reviewed (approved):
As far as I can see, this is fine now

DIzFer commented:
I seem to not understand what this PR is achieving on my system: I'm now getting the intended warning for a `hashedPassword == "!"`... which is exactly what I expect `users.users.root.hashedPasswords` to be. Am I expected (and doomed) to ignore a warning on every rebuild until the end of times?

rnhmjoj commented:
No, you're not supposed to get that warning and this PR will fix it. The warning is intended to check valid hashes but in the original PR (#83171) I missed the edge cases of "!", "*" and others.

nh2 commented:
Confirming that commit 470ce4784e825663dd3357c4e8de07b8012e354f triggers a spurious warning, e.g. for `users.extraUsers.root.password = "";` or `users.extraUsers.root.password = "a";` (i came here via bisection of it).

rnhmjoj commented:
@nh2 Note that It's not `root.password`, which is never checked, but `root.initialHashedPassword` set to "!".
